### ISSUE TITLE
3D tileset in Cesium

### ISSUE THREAD

**Issue by lanminik** [tudelft3d/3dbag-viewer#74]
Hello there,
I am trying to use a 3d tileset (https://3dbag.nl/download/3dtiles/v21031_7425c21b/lod22/tileset.json) and have an issue showing it in Cesium.js. Cesium doesn't report any warning, error, or information and everything seems like it should work but the tiles are not showing at all ( there are no requests for tiles when I am moving around). I tried with another resources that I've found on the internet and tileset.json looks pretty the same as yours, but those worked.

If this requires exporting of your PostgreSQL db in another way to support Cesium, please help (guide) me on how to do that.

BTW, great job!

**Comment by Ylannl** [comment]
Hi,
we are not really offering the 3D tiles as a service for external use at this moment. Our tileset files are optimised for our specific 3D viewer needs, and may not be completely compliant with the 3D tiles standard. That's probably why it doesn't work in Cesium.

If you want to get this working now you can modify the tileset.json yourself (make a flat hierarchy of the child nodes and play with the geometric error). In any case please copy the b3dm files to your own server.

In the future we may be able to make this easier to do.

**Comment by liberostelios** [comment]
The problem is that Cesium uses WGS84 horizontal coordinates, while we use EPSG 28992. You'll have to convert the coordinates of the batched files to make them work with Cesium, as far as I am concerned.

You could try to create your own tileset using [pg2b3dm](https://github.com/Geodan/pg2b3dm) after you load the 3D BAG database in PostGIS and transform it.

**Comment by lanminik** [comment]
Thank you guys for the answers.

I've downloaded all b3dm files and will try somehow to change the projection to fits the Cesium requirements (Plan B). 
In general, it would be much better for me to create a new tileset with pg2b3dm as @liberostelios said but I am not sure what are all steps that I have to do to achieve that.

I've followed the getting started guide and this is what I did since now:
1. Downloaded your PostgreSQL db.
2. Created Postgres server, installed gis extension, and imported the db so now I see pand and other columns.
3. Downloaded bag3d_2020-05-13.gpkg and imported it to the db;
4. Tried to use the tesselate_building library but I got 'height' column is missing (this is the step where I am now)
5. I have found this page [where](http://wiki.bk.tudelft.nl/toi-pedia/Obtaining_data_using_QGIS) is the link for downloading a file with Building footprints (+ height) but I am always redirected to the bag3d viewer (I guess because it is a single-page app and the router redirects me); 

- Is there some other source to get the file?
- Do I go in the right direction to achieve creating a tileset for Cesium or not?

If I don't need some of the steps above please let me know;

Any suggestion or help is much appreciated.

**Comment by DmitriLapushkin** [comment]
I have the same problem. I want to create my own tiles using tool "pg2b3dm" from postgis dump and table "bag3d.lod22_3d" and column "geometrie", where the pure 3d model of the buildings are located in MultiPolygonZ type.

But there is like a dead end. When I`m starting with prepartions. 

My flow:
1. restored postgis 3dbag dump
2. followed the pg2b3dm
3. starting to prepare data for pg2d3dm which requires column of "**lod22_3d.geometrie**" to be in "POLYHEDRALSURFACE Z"
4. getting started guide tells that we need to use this tool https://github.com/bertt/tesselate_building, but it requires **HEIGHT** attribute, we can`t get height because "lod22_3d.geometrie" is already a 3d model

But every doc that I`m reading on http://3dbag.bk.tudelft.nl tells me that for b3dm models pg2b3dm tool was used. But when you start to follow what is required for pg2b3dm, you are stuck. You can`t use bertt/tesselate_building, because it requires hight and you cant get height from lod22_3d.geometrie, because it is alredy a building 3d model.

@Ylannl  Please help us. I think @lanminik is stuck on same thing.

**Comment by Ylannl** [comment]
You shouldn't need to triangulate/tesselate the geometries since they are already triangulated in the DB dump.

**Comment by DmitriLapushkin** [comment]
@Ylannl thanks for clarification! But still there is one piece in puzzle is missing. How we can produce tiles used in http://3dbag.bk.tudelft.nl by using **pg2b3dm** tool? 

Maybe you can give a hint or share a tools you have used to produce your tiles. I suppose you used directly LOD22 models from PostGis database dump (which is available here https://3dbag.nl/en/download) to produce your tilesets. But I dont find anywhere this information.

**Comment by Ylannl** [comment]
We use a modified version of pg2b3dm which has some hacks specific to our viewer that won't work with Cesium. However, the regular pg2b3dm should also work. Just pass it the db parameters.

**Comment by DmitriLapushkin** [comment]
@Ylannl so my assumption was correct about pg2b3dm. But I want it to be used with mapbox. And as i described above, out of the box using pg2b3dm is not possible :)

**Comment by Ylannl** [comment]
You are right we changed it to work with MultiPolygons, see https://github.com/tudelft3d/pg2b3dm/commit/96489b63442f1fdfd28843ce7831d2e8f1d79705.

**Comment by DmitriLapushkin** [comment]
@Ylannl thanks for your help! That is the piece of the puzzle I was looking for. Have not spotted that you have forked this repo. Will try to do the same

**Comment by DmitriLapushkin** [comment]
@Ylannl does it really take several weeks to produce 3d tiles for whole netherlands with pg2b3dm? :) I see you use some sort of precalculated **tables tiles.xyz_tiles** and **tiles.bag_tiles_3k** in time consuming sections. Can you please help, how and what you precompute, cause from source code it is not clear.

**Comment by Ylannl** [comment]
I think you need to create a spatial index in the database on the geometry column. Also we use multithreading in our modified pg2b3dm.

**Comment by DmitriLapushkin** [comment]
@Ylannl yes, thats already creaeted. And you use multithreading only when you output the tiles that can be multithreaded. But not when you do are creating quadtree, since there is recursion.

**Comment by lanminik** [comment]
Yes @DmitriLapushkin we have the same issue with generating the tileset :) Maybe you explained it better.

Did manage it to work? I didn't, so if you are then please give me some steps and hints on how did you achieve it? And did you use the forked pg2b3dm or the original one? Thanks in advance!

**Comment by DmitriLapushkin** [comment]
@lanminik I have archived it but, only they appear in Gulf of Gunea :)
![image](https://user-images.githubusercontent.com/10732799/147738003-5c63b8e9-c76b-4af3-81b8-e409ef5a80b7.png)

And i dont know how to move them to Netherlands, there is some puzzled math involved. One way is to generate them with right SRID for mapbox, which should avoid complex math. But that just takes ages to complete

**Comment by mnHybrid** [comment]
@DmitriLapushkin I am still struggling with this as well. Had the same issue with the positioning in the Gulf of Gunea, and have managed the transformation with query directly in the Postgres DB. Will drop it here when I find it.

Did you manage to generate the geojson data in LOD2.2? (I cannot see from the screenshot); If yes, can you describe the steps that you went through, please? 
I managed somehow to export LOD1 (an extremely slow process) but it is not what I need to have. So have to figure out how to generate lod22.

**Comment by Roempie** [comment]
@Ylannl you say the BAG dataset is already triangulated and it should work out of the box with pg2b3dm. In the meantime I have come a long way by converting the MultiPolygons to PolyhedrasSurfaces, however, the last piece of the puzzle is the amount of vertices present within a polygon. When using pg2b3dm I get this error: (Parameter 'Expected number of vertices in triangles: 4, actual: 28').

This is clearly visible when inspecting a polygon within the BAG db dump, there are 28 vertices present:
SRID=7415;MULTIPOLYGON(((**251826.02832714844 474636.68562890624 20.466999053955078,251826.08002392578 474636.79012109374 20.466999053955078,251826.14447705078 474636.75813867187 20.466999053955078,251830.53931347656 474634.5838222656 20.466999053955078,251830.6320258789 474634.5379238281 20.466999053955078,251830.57587353516 474634.4243984375 20.466999053955078,251827.75531689453 474628.7232265625 20.466999053955078,251827.67999951172 474628.5708828125 20.466999053955078,251827.4805366211 474628.669515625 20.466999053955078,251823.1831123047 474630.79622460937 20.466999053955078,251823.1289741211 474630.8230800781 20.466999053955078,251823.16596142578 474630.89778710937 20.466999053955078,251826.02832714844 474636.68562890624 20.466999053955078**)), ((....))

However, input data coming out of the tesselate_buidling tool only contains 4 vertices per polygon. 
SRID=4978;POLYHEDRALSURFACE(((**3890448.442389822 325755.29241314786 5026791.040034701,3890447.0467412984 325756.262275183 5026792.050525388,3890446.9875947866 325756.1054792064 5026792.1060879165,3890448.442389822 325755.29241314786 5026791.040034701**)),((3890446.9875947866 325756.1054792064 5026792.1060879165,3890446.819034999 325756.24077509344 5026792.226961123,3890446.0132233333 325754.30544179963 5026792.971015101,3890446.9875947866 325756.1054792064 5026792.1060879165)),((3890446.0132233333 325754.30544179963 5026792.971015101,3890445.971589232 325754.2054446789 5026793.009458562,3890446.2087265784 325754.04077940434 5026792.837756099,3890446.0132233333 325754.30544179963 5026792.971015101)),((...)),((...))

Also looking at the forked version of pg2b3dm this same error message is present and thus too requires four vertices per polygon. 

I wonder what approach did you do in order to decrease the amount of vertices present per polygon part to 4 in order to run pg2b3dm on it.

**Comment by rbraggaar** [comment]
Any news on support for Cesium or Mapbox/Deck tile formats?

**Comment by alexhuygen** [comment]
So BAG and Cesium and Unreal can work together?
